File: android/src/main/java/com/lightningkite/khrysalis/observables/CombineObservableProperty.shared.kt
package com.lightningkite.khrysalis.observables

import com.lightningkite.khrysalis.Box
import com.lightningkite.khrysalis.CaptureUnowned
import com.lightningkite.khrysalis.Escaping
import com.lightningkite.khrysalis.boxWrap
import com.lightningkite.khrysalis.rx.combineLatest
import io.reactivex.Observable

class CombineObservableProperty<T, A, B>(
    val observableA: ObservableProperty<A>,
    val observableB: ObservableProperty<B>,
    val combiner: @Escaping() (A, B) -> T
): ObservableProperty<T>() {
    override val value: T
        get() = combiner(observableA.value, observableB.value)
    override val onChange: Observable<Box<T>>
        get() {
            val combinerCopy = combiner
            return observableA.onChange.startWith(Box.wrap(observableA.value))
                .combineLatest(observableB.onChange.startWith(Box.wrap(observableB.value))) @CaptureUnowned("combiner") { a: Box<A>, b: Box<B> -> boxWrap(combinerCopy(a.value, b.value)) }
                .skip(1)
        }
}

fun <T, B, C> ObservableProperty<T>.combine(
    other: ObservableProperty<B>,
    combiner: @Escaping() (T, B) -> C
): ObservableProperty<C> {
    return CombineObservableProperty(this, other, combiner)
}

File: ios/Khrysalis/com/lightningkite/khrysalis/observables/CombineObservableProperty.shared.swift
//Package: com.lightningkite.khrysalis.observables
//Converted using Khrysalis2

import Foundation
import RxSwift
import RxRelay



public class CombineObservableProperty<T, A, B>: ObservableProperty<T> {
    
    public var observableA: ObservableProperty<A>
    public var observableB: ObservableProperty<B>
    public var combiner:  (A, B) -> T
    
    override public var value: T {
        get {
            return combiner(observableA.value, observableB.value)
        }
    }
    override public var onChange: Observable<T> {
        get {
            return observableA.onChange.startWith(observableA.value).combineLatest(observableB.onChange.startWith(observableB.value)){ (a: A, b: B) in
                self.combiner(a.value, b.value)
            }.skip(1)
        }
    }
    
    public init(observableA: ObservableProperty<A>, observableB: ObservableProperty<B>, combiner: @escaping (A, B) -> T) {
        self.observableA = observableA
        self.observableB = observableB
        self.combiner = combiner
        super.init()
    }
    convenience public init(_ observableA: ObservableProperty<A>, _ observableB: ObservableProperty<B>, _ combiner: @escaping (A, B) -> T) {
        self.init(observableA: observableA, observableB: observableB, combiner: combiner)
    }
}
 
 

extension ObservableProperty {
    public func combine<B, C>(other: ObservableProperty<B>, combiner: @escaping (T, B) -> C) -> ObservableProperty<C> {
        return CombineObservableProperty(self, other, combiner)
    }
    public func combine<B, C>(_ other: ObservableProperty<B>, _ combiner: @escaping (T, B) -> C) -> ObservableProperty<C> {
        return combine(other: other, combiner: combiner)
    }
}
 
